firestore/src/main/java/com/firebase/ui/firestore/paging/PageKey.java

```diff
@@ -0,0 +1,48 @@
+package com.firebase.ui.firestore.paging;
+
+import android.support.annotation.Nullable;
+import android.support.annotation.RestrictTo;
+
+import com.google.firebase.firestore.DocumentSnapshot;
+import com.google.firebase.firestore.Query;
+
+/**
+ * Key for Firestore pagination. Holds the DocumentSnapshot(s) that bound the page.
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public class PageKey {
+
+    private final DocumentSnapshot mStartAfter;
+    private final DocumentSnapshot mEndBefore;
+
+    public PageKey(@Nullable DocumentSnapshot startAfter, @Nullable DocumentSnapshot endBefore) {
+        mStartAfter = startAfter;
+        mEndBefore = endBefore;
+    }
+
+    public Query getPageQuery(Query baseQuery, int size) {
+        Query pageQuery = baseQuery;
+
+        if (mStartAfter != null) {
+            pageQuery = pageQuery.startAfter(mStartAfter);
+        }
+
+        if (mEndBefore != null) {
+            pageQuery = pageQuery.endBefore(mEndBefore);
+        } else {
+            pageQuery = pageQuery.limit(size);
+        }
+
+        return pageQuery;
+    }
+
+    @Override
+    public String toString() {
+        String startAfter = mStartAfter == null ? "null" : mStartAfter.getId();
```

Comment:
`String.valueOf(...)`? 😁 Or TBH, I'm pretty sure the compiler does this for you if you just inline everything like so:
```java
        return "PageKey{" +
                "StartAfter=" + mStartAfter.getId() +
                ", EndBefore=" + mEndBefore.getId() +
                '}';
```

Reply:
It's not the ID I am worried about being null, it's `mStartAfter` and `mEndBefore` themselves which can be null. 

In Kotlin I could use `?.` but in Java I have to do this junk.

Reply:
Whoops, yeah, I completely glossed over that one. What you're doing SGTM 👍